### PR TITLE
[DoctrineBridge] Deprecated using IdReader when optimization is not possible

### PR DESCRIPTION
**Pull request by HeahDude**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | yes 
| Tests pass?   | yes    
| Fixed tickets | ~
| License       | MIT
| Doc PR        | ~ 

Follow up of #30962. (Review only the second commit until #30962 is merged).

### CODE REVIEW

**Comment by fabpot** [comment]
@HeahDude  #30962 has been merged now.

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php [hunk 1]
```diff
@@ -49,6 +49,10 @@ public function __construct(ObjectManager $manager, string $class, IdReader $idR
     {
         $classMetadata = $manager->getClassMetadata($class);
 
+        if ($idReader && !$idReader->isSingleId()) {
+            @trigger_error(sprintf('Passing an instance of "%s" with an entity class "%s" that have a composite id is deprecated since 4.3 and will throw an exception in 5.0.', IdReader::class, $class), E_USER_DEPRECATED);
```
Comment (yceruto) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: I don't get this, how do you know before hand that the entity has composite id without check `$idReader->isSingleId()`?

Comment (HeahDude) [reply src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: See https://github.com/symfony/symfony/pull/30962 description.

File: src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php [hunk 2]
```diff
@@ -42,13 +42,17 @@ class DoctrineChoiceLoader implements ChoiceLoaderInterface
...
-     * @param IdReader                   $idReader     The reader for the object IDs
+     * @param IdReader|null              $idReader     The reader for the object IDs
      * @param EntityLoaderInterface|null $objectLoader The objects loader
      */
     public function __construct(ObjectManager $manager, string $class, IdReader $idReader = null, EntityLoaderInterface $objectLoader = null)
     {
         $classMetadata = $manager->getClassMetadata($class);
 
+        if ($idReader && !$idReader->isSingleId()) {
+            @trigger_error(sprintf('Passing an instance of "%s" with an entity class "%s" that have a composite id is deprecated since 4.3 and will throw an exception in 5.0.', IdReader::class, $class), E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: [...] that has a [...]

Comment (xabbuh) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: [...] since Symfony 4.3 [...]

Comment (xabbuh) [inline src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php]: We should also add a commented line of PHP code that would just need to be enabled on 5.0.

**Review by xabbuh — CHANGES_REQUESTED** [review]
This is missing entries in the `UPGRADE-4.3.md` and `UPGRADE-5.0.md` files.

File: src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php [hunk 3]
```diff
@@ -240,7 +234,11 @@ public function configureOptions(OptionsResolver $resolver)
                 $this->idReaders[$hash] = new IdReader($options['em'], $classMetadata);
             }
 
-            return $this->idReaders[$hash];
+            if ($this->idReaders[$hash]->isSingleId()) {
+                return $this->idReaders[$hash];
+            }
+
+            return null;
```
Comment (yceruto) [inline src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php]: Ha, this is exactly what I was trying to explain in our discussion :) good

Comment (HeahDude) [reply src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php]: Yes this was implemented differently and lost in a rebase :)

**Review by xabbuh — APPROVED** [review]
with the class clarification in the deprecation message we just discussed

**Comment by HeahDude** [comment]
done :), thanks!

File: src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php [hunk 4]
```diff
@@ -181,12 +178,9 @@ public function configureOptions(OptionsResolver $resolver)
         // are indexed by an incrementing integer.
         // Use the ID/incrementing integer as choice value.
         $choiceValue = function (Options $options) {
-            /** @var IdReader $idReader */
-            $idReader = $options['id_reader'];
-
             // If the entity has a single-column ID, use that ID as value
-            if ($idReader->isSingleId()) {
-                return [$idReader, 'getIdValue'];
+            if ($options['id_reader'] instanceof IdReader && $options['id_reader']->isSingleId()) {
```
Comment (yceruto) [inline src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php]: (optional) If we trust the normalizer, maybe we can get rid of `instanceof` everywhere and compare with` null`.

**Review by yceruto — APPROVED** [review]

File: src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php (hunk 4 above)
Comment (HeahDude) [reply src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php]: It helps autocomplete while getting rid of the duck type hinting I've removed in the process.

**Review by stof — APPROVED** [review]

**Comment by stof** [comment]
Thank you @HeahDude.

**Pull request merged**